### ISSUE TITLE
Add GPU-deterministic back-prop for fused softmax/cross-entropy ops

### ISSUE THREAD

**Issue by duncanriach** [tensorflow/tensorflow#38185]
**SYSTEM INFORMATION**

- TensorFlow version (you are using): `2.2.0-rc2`
- Are you willing to contribute it: Yes (please assign it to me)

**CURRENT BEHAVIOR**

The back-prop of `tf.nn.softmax_cross_entropy_with_logits` and `tf.nn.sparse_softmax_cross_entropy_with_logits` is non-deterministic on GPUs.

**WILL THIS CHANGE THE CURRENT API?**

No. Assuming the deterministic back-prop kernels are slower than the current non-deterministic ones, then the deterministic operation will be selectable using the preferred mechanism at the time. At the time of writing, that mechanism is to set the environment variable `TF_DETERMINISTIC_OPS` to "1" or "true".

**WHO WILL BENEFIT FROM THIS FEATURE?**

Determinism, for both training and inference, is becoming increasingly important as deep learning systems are moved into production, not only because of regulatory requirements of some markets but also, more broadly, because of the massive potential performance advantages of training with determinism; determinism accelerates and facilitates debug, experimentation (including hyper-parameter tuning and active learning), and regression testing.

With `TF_DETERMINISTIC_OPS` now enabling deterministic functionality of cuDNN convolution, bias addition, max-pooling, and CTC loss, many deep learning models will train deterministically on GPUs. Softmax and cross entropy are both foundational functions in deep learning, and are combined [to ensure performance and numerical stability](https://github.com/tensorflow/tensorflow/issues/2462
). Enabling these fused ops to function deterministically will enhance the ability for TensorFlow to be used for various production systems.

This current issue comes out of [Issue 9](https://github.com/NVIDIA/tensorflow-determinism/issues/9) in the [tensorflow-determinism](https://github.com/NVIDIA/tensorflow-determinism) project, which is focused on making the `SequenceToSequece` model of [OpenNMT-tf](https://github.com/OpenNMT/OpenNMT-tf) train deterministically.

**UNIT TESTS**

What follows are essentially production-ready TensorFlow unit tests that currently fail, but that will pass when this feature is implemented correctly. Depending on the implementation, the parameter-space covered by the tests (e.g. "batch" dimensionality) may need to be broadened for coverage. The tests can be seen running (and failing) on TensorFlow version `2.2.0-rc2` in [this colab](https://colab.research.google.com/drive/1syj32Jl7dS6mBa-GhNrq_LLIxvfumIOz).

```
import tensorflow as tf
import numpy as np

class DeterministicTest(tf.test.TestCase):

  def _randomInts(self, shape, high, dtype):
    return tf.constant(
        np.random.randint(low=0, high=high, size=shape).astype(dtype))

  def _randomFloats(self, shape, dtype, normalized_rows=False):
    a = (2 * np.random.random_sample(shape) - 1).astype(dtype)

    if normalized_rows:

      def normalize(row):
        return row / row.sum()

      a = np.apply_along_axis(normalize, 1, a)

    return tf.constant(a)
    
  def _testDeterministicGradients(self, exclusive_labels):
    with self.session(force_gpu=True):
      batch_size = 1024
      classes_count = 1000
      logits_shape = (batch_size, classes_count)
      logits_dtype = np.float32
      logits = self._randomFloats(logits_shape, logits_dtype)
      if exclusive_labels:
        labels_shape = (batch_size)
        labels_dtype = np.int32
        labels = self._randomInts(labels_shape, classes_count, labels_dtype)
      else:
        labels_shape = logits_shape
        labels_dtype = logits_dtype
        labels = self._randomFloats(labels_shape, labels_dtype,
                                    normalized_rows=True)
      output_shape = (batch_size)
      output_dtype = logits_dtype

      def gradients(local_seed):
        np.random.seed(local_seed)
        upstream_gradients = self._randomFloats(output_shape, output_dtype)
        with tf.GradientTape(persistent=True) as tape:
          tape.watch(logits)
          if exclusive_labels:
            tested_op = tf.nn.sparse_softmax_cross_entropy_with_logits
          else:
            tested_op = tf.nn.softmax_cross_entropy_with_logits
          op_output = tested_op(labels=labels, logits=logits)
          gradient_injector_output = op_output * upstream_gradients
        return tape.gradient(gradient_injector_output, logits)

      repeat_count = 5
      for seed in range(repeat_count):
        result_a = gradients(seed)
        result_b = gradients(seed)
        self.assertAllEqual(result_a, result_b)

  def testExclusiveLabelsDeterministicGradients(self):
    self._testDeterministicGradients(exclusive_labels=True)

  def testDistributionLabelsDeterministicGradients(self):
    self._testDeterministicGradients(exclusive_labels=False)

if __name__ == '__main__':
  tf.test.main()
```

**POSSIBLE SOLUTION APPROACHES**

The ultimate solution will be to fix this at the CUDA level by creating deterministic kernels that are exactly functionally equivalent to the existing non-deterministic kernels. I'll soon be upstreaming a solution like this for `tf.image.resize` with `method=ResizeMethod.BILINEAR`; see [PR 39243](https://github.com/tensorflow/tensorflow/pull/39243).

It could also be possible to create a temporary solution at the Python level by combining existing deterministic ops. See deterministic [tf.nn.bias_add](https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/ops/nn_ops.py#L2741-L2755) as an example. Assuming that `tf.nn.softmax` operates deterministically (I have not tested that), then this could be followed by an arithmetically robust implementation of cross-entropy (e.g. [streaming log-sum-exp](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html)). There would, of course, be a performance cost (1/2 speed) compared to a fused softmax and cross-entopy.

**ASSIGNMENT**

I'm willing and able to implement this feature, but I might not be able to get to it for quite a while. I'm happy for someone else to implement it, but please let me know if you start actively working on it (and I will do the same), so that we can avoid duplicating effort.

**Comment by duncanriach** [comment]
I just added two sections to my original comment: POSSIBLE SOLUTION APPROACHES and ASSIGNMENT.

**Comment by duncanriach** [comment]
Work-around info: use non-fused softmax and cross-entropy. For example, select the activation on the final `Dense` keras layer to be `softmax` and then select `tf.keras.losses.categorical_crossentropy` for the loss function.

**Comment by MFreidank** [comment]
@duncanriach Thanks for helping me resolve my issue (yesterday on the `tensorflow-determinism` repo) that revolved around this very problem. 

The (cleanest) CUDA solution approach you outline is probably out of scope for me. 
However, I believe I am able (and would be happy to) implement a temporary python-side solution.

Checking some assumptions before jumping out there and starting work on this:
* would you be fine with me taking this on?
* this would essentially be a special case in the body of `tf.nn.softmax_cross_entropy_with_logits` and `tf.nn.sparse_softmax_cross_entropy_with_logits` that becomes active if and only if `TF_DETERMINISTIC_OPS` environment variable is set?
* to work around this, it should be possible to follow what `tf.keras` does in the non-fused case you mention as workaround above. Seems to come down to computing `tf.nn.softmax` and then following implementation of `tf.keras.backend.categorical_crossentropy(from_logits=False)` ([lines 4636 - 4641](https://github.com/tensorflow/tensorflow/blob/d5b3ec27d1d6bb157588ff3033a3d9bd2e46711f/tensorflow/python/keras/backend.py#L4636)) Does that sound correct? I'll make sure functionality is retained as per the unit tests.

**Comment by duncanriach** [comment]
> would you be fine with me taking this on?

Absolutely. I expect that you'll do a great job of it.

> this would essentially be a special case in the body of `tf.nn.softmax_cross_entropy_with_logits` and `tf.nn.sparse_softmax_cross_entropy_with_logits` that is active if and only if `TF_DETERMINISTIC_OPS` environment variable is set?

Yes, please refer to how I did this for [`tf.nn.bias_add`](https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/ops/nn_ops.py#L2741-L2755) and follow that patten.

> Seems to come down to computing `tf.nn.softmax`.

I think you can/should just use `tf.nn.softmax` directly.

> and then following implementation of tf.keras.backend.categorical_crossentropy(from_logits=False)

I think you can/should just use `tf.keras.losses.categorical_crossentropy(from_logits=False)` and `tf.keras.losses.sparse_categorical_crossentropy(from_logits=False)` directly also. I don't fully remember what the issue with arithmetic robustness was (and the need to consider [streaming log-sum-exp](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html)). That might only be something to consider when implementing the final CUDA solution.

> I'll make sure functionality is retained as per the unit tests.

Please use the test code in my original comment on this issue.

Reference how I organized the determinism tests for `tf.nn.bias_add` in [tensorflow/python/kernel_tests/](https://github.com/tensorflow/tensorflow/tree/d5b3ec27d1d6bb157588ff3033a3d9bd2e46711f/tensorflow/python/kernel_tests): the existing test moved into [bias_op_base.py](https://github.com/tensorflow/tensorflow/blob/d5b3ec27d1d6bb157588ff3033a3d9bd2e46711f/tensorflow/python/kernel_tests/bias_op_base.py) and then included in [bias_op_test.py](https://github.com/tensorflow/tensorflow/blob/d5b3ec27d1d6bb157588ff3033a3d9bd2e46711f/tensorflow/python/kernel_tests/bias_op_test.py) and then the deterministic test cases add in [bias_op_deterministic_test.py](https://github.com/tensorflow/tensorflow/blob/d5b3ec27d1d6bb157588ff3033a3d9bd2e46711f/tensorflow/python/kernel_tests/bias_op_deterministic_test.py).

You'll need to find the existing op test and where the gradient is tests (sometimes it's in a different file) and then appropriately add in the deterministic tests.

Review the changes in [PR 31465](https://github.com/tensorflow/tensorflow/pull/31465), including the change to `tensorflow/tools/pip_package/BUILD`.

---------

Another option is for you to begin by adding a patch to tensorflow-determinism (applied with tfd.patch()). This could be released almost immediately (in version 0.4) for application to stock TensorFlow 2.2 (and even earlier versions). Then either you or I could take that solution and upstream it into stock TensorFlow for release in version 2.3 or 2.4 (depending on how long the review process takes).

The advantage of this approach is that it allows almost immediate release and utility along with the ability to focus on getting the implementation correct before dealing with the process of getting it correctly integrated into, and released via, stock TensorFlow.

**Comment by MFreidank** [comment]
@duncanriach I like the suggested option to start from `tensorflow-determinism` as it would get this shipped to users at earliest. 
I'll start on it there and once we have confidence the fix works upstreaming it should be much more straightforward.

Thanks for the guidance, really appreciate it!

**Comment by duncanriach** [comment]
Note that [PR 47925](https://github.com/tensorflow/tensorflow/pull/47925), which has been merged and will be present in TensorFlow version 2.6 onwards, causes TensorFlow to throw `tf.errors.UnimplementedError` when `TF_DETERMINISTIC_OPS` is set to `"true"` or `"1"` (when determinism is expected) and the softmax/cross-entropy ops are used in such a way that truly random noise will be injected into the model's functionality.

This exception-throwing functionality can currently be disabled by setting `TF_DISABLE_SPARSE_SOFTMAX_XENT_WITH_LOGITS_OP_DETERMINISM_EXCEPTIONS` to `"true"` or `"1"` in case you're using a work-around (such as the one [described here](https://github.com/NVIDIA/framework-determinism/blob/master/tensorflow_status.md#fused-softmaxcross-entropy)) that relies on the existing GPU-op functionality being exposed and usable.

This move to implement d9m-unimplemented exception throwing is orchestrated under this [determinism RFC](https://github.com/tensorflow/community/blob/master/rfcs/20210119-determinism.md).

**Comment by duncanriach** [comment]
Note that PR [50070](https://github.com/tensorflow/tensorflow/pull/50070) has been merged and will be present in TF 2.7.  Deterministic functionality for both `tf.nn.softmax_cross_entropy_with_logits` and `tf.nn.sparse_softmax_cross_entropy_with_logits` (and the Keras layers that utilize them) can be enabled using `tf.config.experimental.enable_deterministic_ops(True)`.
